lib/cli-engine/config-array-factory.js

```diff
@@ -697,6 +697,33 @@ class ConfigArrayFactory {
             ctx.matchBasePath
         );
 
+        /**
+         * Cloning the rule's config as we are setting `useDefaults` to true`
+         * which mutates the config with its default value if present. And when we
+         * refer to a same variable for config for different rules, that referred variable will
+         * be mutated and it will be used for both.
+         *
+         * Example:
+         *
+         * const commonRuleConfig  = ['error', {}];
+         *
+         * Now if we use this variable as a config for rules like this
+         *
+         * {
+         *      rules: {
+         *          "a" : commonRuleConfig,
+         *          "b" : commonRuleConfig,
+         *      }
+         * }
+         *
+         * And if these rules have default values in their schema, their
+         * config will be mutated with default values, the mutated `commonRuleConfig` will be used for `b` as well and it probably
+         * throw schema voilation errors.
+         *
+         * Refer https://github.com/eslint/eslint/issues/12592
+         */
+        const clonedRulesConfig = rules && JSON.parse(JSON.stringify((rules)));
```

Comment:
unfortunately i never saw this PR; JSON cloning should never be used, especially since eslint config values can be regexes, Infinity, NaN, and other things that don't survive JSON serialization.

Reply:
@ljharb We did tried to avoid json cloning using this. But we tried using lodash cloneDeepWith with customizer but that had different issues like circular ref issue,

Reply:
This is a BREAKING CHANGE for a lot of plugins which support options via

> regexes, Infinity, NaN, and other things that don't survive JSON serialization

For example: https://github.com/sindresorhus/eslint-plugin-unicorn/blob/master/docs/rules/filename-case.md#ignore